# Optimisasi Matematika Dengan Python
### dan beberapa aplikasinya di bidang engineering dan teknik kimia
Zulfan Adi Putra  
THE Chemical Engineer


<a id='00_Daftar_Isi'></a>
### Daftar Isi
01. [Optimisasi Matematika Teknik Kimia](#01_Optimisasi_Matematika)
02. [Unconstrained Optimization](#02_Unconstrained_Optimization)
03. [Constrained Optimization](#03_Constrained_Optimization)
04. [Linear Programming](#04_Linear_Programming)
05. [Integer Programming](#05_Integer_Programming)
06. [Nonlinear Optimization](#06_Nonlinear_Programming)
07. [Stochastic Optimization](#07_Stochastic_Programming)
08. [Robust Optimization](#08_Robust_Programming)   
09. [Dynamic Optimization](#09_Dynamic_Optimization)
    



### Literature
- 
- 
- Mitsos, A., 2021, Mathematical Optimization for Engineers EDX, RWTH Aachen   
- Practical Optimization: a Gentle Introduction, in https://www.optimization101.org/.   
- Biegler, L.T., 2021, Nonlinear Optimization in Process Systems Engineering, in http://numero.cheme.cmu.edu/course/06606.html  



<a id='01_Optimisasi_Matematika'></a>
### 01. Optimisasi Matematika di bidang Teknik Kimia

Optimisasi matematika adalah proses pemodelan suatu masalah, baik masalah terkait engineering, bisnis, atau apapun, ke dalam persamaan/pertidaksamaan matematika Salah satu contohnya adalah mencari nilai akar suatu persamaan kuadrat.  
Secara umum, optimisasi matematika diformulasikan sebagai berikut:  

Fungsi objektif yang akan dicari nilai minimumnya:  
$min f(x,y,z)$  

Terikat kepada *constraint* di bawah ini:  
$h_i(x,y,z) = 0$  

$g_j(x,y,z) \le 0$   

$x \in R$  

$y \in \{0,1\}$  

$z \in I$   


Di bab ini kita akan belajar beberapa contoh bagaimana menggunakan teknik-teknik optimisasi yang umum untuk menyelesaikan permasalahan di bidang teknik kimia. Beberapa contoh terkait bidang _operation research_ atau _systems engineering_ juga akan dipakai di sini.

<a id='02_Unconstrained_Optimization'></a>
### 02. Unconstrained optimization
Unconstrained optimization berarti optimisasi suatu fungsi objektif tanpa adanya *constraint* atau halangan. Salah satu contoh sederhana adalah mencari nilai akar suatu persamaan kuadratik. Dengan menggunakan metode Newton, kita mengetahui bahwa nilai maksimum atau minimum akan didapat jika turunan pertama suatu fungsi kuadrat sama dengan nol.   

Kita asumsikan fungsi sederhana $ y = x^2$  
Tentunya kita telah mengetahui bahwa pada nilai $x = 0$, maka nilai $y$ pun bernilai minimum sebesar $0$ juga. 

In [53]:
# definisikan fungsinya
def y(x):
    return x**2

In [54]:
# selesaikan dengan menggunakan scipy.optimize.root
from scipy.optimize import root

# kita tetapkan nilai awalnya
x0 = 10
hasil = root(y, x0)
print('Nilai y sebesar: ', hasil.fun)
print('Nilai x yang menghasilkan nilai y tersebut adalah: ',hasil.x)

Nilai y sebesar:  [3.22743728e-165]
Nilai x yang menghasilkan nilai y tersebut adalah:  [5.68105384e-83]


Paket lain untuk nonlinear unconstrained optimization adalah dengan menggunakan _scipy.optimize.minimize_. Contohnya adalah sebagai berikut.

In [57]:
import numpy as np
from scipy.optimize import minimize

# fungsi yang mau dicari nilai minimumnya
def rosen(x):
    rosen = sum(100*(x[1:] - x[:-1]**2)**2 + (1-x[:-1])**2)
    return rosen

# nilai awal
x0 = np.array([1.3, 0.7])

hasil = minimize(rosen, x0, method='nelder-mead', options={'xatol': 1e-8, 'disp': True})
print(hasil)
print()
print('Nilai minimum fungsi rosen adalah:', hasil.fun)
print('Nilai x yang memberikan fungsi rosen minimum adalah:', hasil.x)

Optimization terminated successfully.
         Current function value: 0.000000
         Iterations: 79
         Function evaluations: 150
 final_simplex: (array([[1.        , 1.        ],
       [1.        , 0.99999999],
       [1.        , 1.        ]]), array([3.37360776e-18, 9.41642436e-18, 2.25070018e-17]))
           fun: 3.3736077629532093e-18
       message: 'Optimization terminated successfully.'
          nfev: 150
           nit: 79
        status: 0
       success: True
             x: array([1., 1.])

Nilai minimum fungsi rosen adalah: 3.3736077629532093e-18
Nilai x yang memberikan fungsi rosen minimum adalah: [1. 1.]


Nah, sekarang silahkan cari fungsi-fungsi lainnya yang akan dicari nilai minimum atau maksimumnya. 

[Kembali ke daftar isi](#00_Daftar_Isi)

<a id='03_Constrained_Optimization'></a>
### 03. Constrained optimization
Constrained optimization berkebalikan daripada unconstrained optimization. Pada kenyataannya, hampir semua permasalahan optimisasi akan memiliki *constraint* atau halangan.  

Contoh berikut ini adalah jika nilai variabel pengambil keputusannya ($x$) berada pada rentang tertentu.  

In [63]:
# impor paket
import numpy as np
import pandas as pd
from scipy.optimize import minimize_scalar, minimize, Bounds

# definisikan fungsi objektif
def objective(x):
    LVGO_PA_Flow = x[0]
    LVGO_PA_Temp = x[1]
    MPA_PA_Flow = x[2]
    MPA_PA_Temp = x[3]
    V3SS_PA_Flow = x[4]
    V3SS_PA_Temp = x[5]
    VR_PA_Flow = x[6]
    VR_PA_Temp = x[7]
    
    obj = 71223.245 + 0.09981*LVGO_PA_Flow - 94.359*LVGO_PA_Temp \
            - 138.5806*MPA_PA_Temp - 108.8623*V3SS_PA_Temp \
            + (LVGO_PA_Flow - 12500.0)*(LVGO_PA_Temp - 55.0)*(-0.001654) \
            + (LVGO_PA_Flow - 12500.0)*(MPA_PA_Flow - 72500.0)*(-4.81e-7) \
            + (LVGO_PA_Flow - 12500.0)*(MPA_PA_Temp - 265.0)*(-0.0019248) \
            + (LVGO_PA_Flow - 12500.0)*(V3SS_PA_Flow - 65000.0)*(-6.90907e-7) \
            + (LVGO_PA_Flow - 12500.0)*(V3SS_PA_Temp - 325.0)*(-0.001514) \
            + (LVGO_PA_Flow - 12500.0)*(VR_PA_Temp - 327.5)*(-0.0007178) \
            + (LVGO_PA_Temp - 55.0)*(MPA_PA_Temp - 265.0)*(1.91343) \
            + (LVGO_PA_Temp - 55.0)*(V3SS_PA_Temp - 325.0)*(1.61218) \
            + (MPA_PA_Flow - 72500.0)*(MPA_PA_Temp - 265.0)*(-0.0013) \
            + (MPA_PA_Temp - 265.0)*(V3SS_PA_Temp - 325.0)*(2.7763) \
            + (V3SS_PA_Flow - 65000.0)*(V3SS_PA_Temp - 325.0)*(-0.00107) \
            + (V3SS_PA_Temp - 325.0)*(VR_PA_Temp - 327.5)*(-0.40576) \
    
    return obj

In [64]:
# data awal
LVGO_PA_Flow = 100000.0
LVGO_PA_Temp = 65.0
MPA_PA_Flow = 110000.0
MPA_PA_Temp = 290.0
V3SS_PA_Flow = 40000.0
V3SS_PA_Temp = 290.0
VR_PA_Flow = 8000.0
VR_PA_Temp = 355.0

x0 = [LVGO_PA_Flow, LVGO_PA_Temp,
     MPA_PA_Flow, MPA_PA_Temp,
     V3SS_PA_Flow, V3SS_PA_Temp,
     VR_PA_Flow, VR_PA_Temp
     ]

In [65]:
# fungsi objektif saat x0
print(objective(x0))
print()

-3782.4084375000175



In [67]:
# batas nilai x
bounds = ((100000.0, 150000.0),
         (45.0, 65.0),
         (35000.0, 110000.0),
         (240.0, 290.0),
         (40000.0, 90000.0),
         (290.0, 360.0),
         (8000.0, 26000.0),
         (300.0, 355.0))

hasil = minimize(objective, x0, method='trust-constr', bounds=bounds, 
                 options={'maxiter': 2000})

print('Kondisi optimisasi adalah:', hasil.message)
print()
print('Nilai minimum fungsi objektifnya adalah: ', hasil.fun)
print()
print('Nilai x optimum adalah:', hasil.x)

Kondisi optimisasi adalah: The maximum number of function evaluations is exceeded.

Nilai minimum fungsi objektifnya adalah:  -18766.044954429442

Nilai x optimum adalah: [1.25349135e+05 6.49998870e+01 1.09999808e+05 2.89999931e+02
 7.63598425e+04 3.59999908e+02 2.59999021e+04 3.54999790e+02]


Kita juga bisa menggunakan solver stochastic programming yang tersedia di _scipy_ seperti:  
- shgo, 
- dual_annealing, 
- differential_evolution, 
- basinhopping

In [68]:
from scipy import optimize
hasil_shgo = optimize.shgo(objective, bounds)
print(hasil_shgo)

     fun: -22137.25831250002
 message: 'Failed to find a feasible minimiser point. Lowest sampling point = -22137.25831250002'
    nfev: 257
     nit: 2
   nlfev: 0
   nlhev: 0
   nljev: 0
 success: False
       x: array([1.50e+05, 6.50e+01, 1.10e+05, 2.90e+02, 9.00e+04, 3.60e+02,
       2.60e+04, 3.55e+02])


In [69]:
hasil_dualannealing = optimize.dual_annealing(objective, bounds)
print(hasil_dualannealing)

     fun: -22137.25831250002
 message: ['Maximum number of iteration reached']
    nfev: 16325
    nhev: 0
     nit: 1000
    njev: 0
  status: 0
 success: True
       x: array([1.5000000e+05, 6.5000000e+01, 1.1000000e+05, 2.9000000e+02,
       9.0000000e+04, 3.6000000e+02, 8.7189795e+03, 3.5500000e+02])


In [70]:
hasil_de = optimize.differential_evolution(objective, bounds)
print(hasil_de)

     fun: -22137.25831250002
     jac: array([-7.23957783e-02, -2.17521301e+02, -9.85892257e-02, -3.35686127e+02,
       -1.32422429e-01, -2.69417069e+02,  0.00000000e+00, -1.12898852e+02])
 message: 'Optimization terminated successfully.'
    nfev: 8736
     nit: 70
 success: True
       x: array([1.50000000e+05, 6.50000000e+01, 1.10000000e+05, 2.90000000e+02,
       9.00000000e+04, 3.60000000e+02, 1.47970512e+04, 3.55000000e+02])


In [71]:
hasil_basinhopping = optimize.basinhopping(objective, bounds)
print(hasil_basinhopping)

                        fun: -1.6468913758157021e+19
 lowest_optimization_result:       fun: -1.6468913758157021e+19
 hess_inv: array([[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 

Nah, sekarang silahkan cari fungsi-fungsi lainnya yang akan dicari nilai minimum atau maksimumnya.

[Kembali ke daftar isi](#00_Daftar_Isi)

<a id='04_Linear_Programming'></a>
### 04. Linear programming
Linear Programming (LP) adalah metode untuk menyelesaikan permasalahan optimisasi matematika di mana fungsi objektif dan semua *constraint* nya berupa persamaan atau pertidaksamaan linear.  

Sebagai contoh, katakanlah sebuah pabrik memproduksi 4 jenis produk yang jumlah masing-masingnya tidak diketahui. Kita simbolkan jumlah produk 1, 2, 3, dan 4 sebagai $x_1, x_2, x_3, x_4$, secara berurutan.  Beberapa kondisi operasi yang harus dipenuhi adalah:
- Keuntungan per unit produk sebesar $ \$20, \$12, \$40 $, dan $\$25 $ untuk produk 1, 2, 3, 4, secara berurutan,
- Jumlah keseluruhan produk yang dihasilkan perhari tidak bisa melebihi 50 unit karena keterbatasan tenaga kerja,
- 1 unit Produk 1 menggunakan 3 unit bahan baku A,
- 1 unit Produk 2 menggunakan 2 unit bahan baku A dan 1 unit bahan baku B,
- 1 unit Produk 3 menggunakan 1 unit bahan baku A dan 2 unit bahan baku B,
- 1 unit Produk 4 menggunakan 3 unit bahan baku B,
- Dalam sehari, pabrik cuma bisa menggunakan 100 unit bahan baku A dan 90 unit bahan baku B.  

Yang ditanyakan adalah berapa unit yang harus diproduksi untuk masing-masing Produk 1, 2, 3, dan 4 sehingga keuntungan pabrik maksimal?

Secara matematika, permasalahan ini dapat ditulis sebagai berikut:  
maksimumkan keuntungan sebagai fungsi objektif = $20x_1 + 12x_2 + 40x_3 +25x_4 $  

*Constraint*nya adalah:  
$x_1 + x_2 + x_3 + x_4 \le 50$ --> untuk keterbatasan tenaga kerja  
$3x_1 + 2x_2 + x_3 \le 100$ --> untuk keterbatasan bahan baku A  
$x_2 + 2x_3 + 3x_4 \le 90$ --> untuk keterbatasan bahan baku B  
$x_1, x_2, x_3, x_4 \ge 0$ --> setiap unit harus bernilai positif atau setidaknya nol.

Oleh karena LP banyak digunakan di dalam kehidupan sehari-hari, maka bab ini dikembangkan dan dibagi lagi menjadi beberapa sub-bab tentang beberapa paket yang terdapat di Python dan beberapa aplikasinya. Di bawah ini adalah tiga modul yang umum digunakan untuk melakukan optimisasi dengan menggunakan Python:  
- [Menggunakan scipy.linprog](#04_01_scipy_linprog)
- [Menggunakan gekko](#04_02_gekko)
- [Menggunakan pyomo](#04_03_pyomo)


Aplikasi-aplikasi di bawah ini dibuat dengan menggunakan Pyomo:   
- [Optimisasi jaringan atau network optimization](#04_03_01_network_optimization)
- [Pencampuran linier atau linear blending problem](#04_03_02_linear_blending)
- [Masalah pengalokasian atau assignment](#04_03_03_assignment)
- [Pemilihan proyek](#04_03_04_project_selection)
- [Piecewise linearization](#04_03_05_piecewise_linearization)


<a id='04_01_scipy_linprog'></a>
#### 04.01 Menggunakan scipy.linprog
Untuk menyelesaikan permasalahan di atas, paket scipy.linprog dapat digunakan sebagai berikut:

In [55]:
# impor paket
from scipy.optimize import linprog

# buat matriks fungsi objektif
obj = [-20, -12, -40, -25]

# buat matriks constraints bagian A dan b di Ax <= b
A = [[1,1,1,1], #tenaga kerja
    [3,2,1,0], #material A
    [0,1,2,3]] #material B

b = [50, #tenaga kerja
    100, #material A
    90] #material B

# rumuskan permasalahannya. 
# The method can be interior point, simplex, or revised simplex
opt = linprog(c=obj, A_ub=A, b_ub=b, method='simplex')
print(opt)

     con: array([], dtype=float64)
     fun: -1899.9999999999998
 message: 'Optimization terminated successfully.'
     nit: 4
   slack: array([0.00000000e+00, 4.00000000e+01, 1.42108547e-14])
  status: 0
 success: True
       x: array([ 5.,  0., 45.,  0.])


<a id='04_02_gekko'></a>
#### 04.02 Menggunakan gekko

In [19]:
# impor paket
from gekko import GEKKO

# buat model
m = GEKKO()

# definisikan variabel bebasnya
x1 = m.Var(lb=0)
x2 = m.Var(lb=0)
x3 = m.Var(lb=0)
x4 = m.Var(lb=0)

# modelkan fungsi objektifnya
# di sini, fungsi objektifnya dibuat negatif karena by default, gekko adalah minimization
m.Obj(-20*x1-12*x2-40*x3-25*x4)

# tambahkan constraintnya
m.Equation(x1+x2+x3+x4 <= 50)
m.Equation(3*x1+2*x2+x3 <= 100)
m.Equation(x2+2*x3+3*x4 <= 90)

In [20]:
# jalankan optimisasinya
m.solve()

apm 161.142.148.110_gk_model2 <br><pre> ----------------------------------------------------------------
 APMonitor, Version 1.0.1
 APMonitor Optimization Suite
 ----------------------------------------------------------------
 
 
 
 --------- APM Model Size ------------
 Each time step contains
   Objects      :            0
   Constants    :            0
   Variables    :            7
   Intermediates:            0
   Connections  :            0
   Equations    :            4
   Residuals    :            4
 
 Number of state variables:              7
 Number of total equations: -            3
 Number of slack variables: -            3
 ---------------------------------------
 Degrees of freedom       :              1
 
 **********************************************
 Steady State Optimization with Interior Point Solver
 **********************************************
  
  
 Info: Exact Hessian

******************************************************************************
This program

In [21]:
# print hasilnya
print("Hasil optimalnya adalah:")
print("x1 =", x1.value[0])
print("x2 =", x2.value[0])
print("x3 =", x3.value[0])
print("x4 =", x4.value[0])
print("Nilai optimumnya adalah ", m.options.objfcnval)

Hasil optimalnya adalah:
x1 = 5.0000000046
x2 = 0.0
x3 = 45.000000023
x4 = 0.0
Nilai optimumnya adalah  -1900.000001


Hasil ini sama dengan yang didapatkan lewat scipy.linprog

<a id='04_03_pyomo'></a>
#### 04.03 Menggunakan pyomo

Dengan Pyomo kita akan mendefinisikan dua jenis model, yakni model concrete dan model abstrak. 

In [41]:
from pyomo.environ import *

In [46]:
# untuk model concrete
model = ConcreteModel()

# definisikan variabel bebas
model.x1 = Var(within=NonNegativeReals)
model.x2 = Var(within=NonNegativeReals)
model.x3 = Var(within=NonNegativeReals)
model.x4 = Var(within=NonNegativeReals)

# definisikan fungsi objektif
model.objective = Objective(
                    expr = 20*model.x1+12*model.x2+40*model.x3+25*model.x4,
                    sense = maximize)

# definisikan constraint
model.constraint1 = Constraint(expr = model.x1+model.x2+model.x3+model.x4 <= 50)
model.constraint2 = Constraint(expr = 3*model.x1+2*model.x2+model.x3 <= 100)
model.constraint3 = Constraint(expr = model.x2+2*model.x3+3*model.x4 <= 90)

In [47]:
# jalankan optimisasinya
solver = SolverFactory('glpk')
solver.solve(model)

{'Problem': [{'Name': 'unknown', 'Lower bound': 1900.0, 'Upper bound': 1900.0, 'Number of objectives': 1, 'Number of constraints': 4, 'Number of variables': 5, 'Number of nonzeros': 11, 'Sense': 'maximize'}], 'Solver': [{'Status': 'ok', 'Termination condition': 'optimal', 'Statistics': {'Branch and bound': {'Number of bounded subproblems': 0, 'Number of created subproblems': 0}}, 'Error rc': 0, 'Time': 0.041345834732055664}], 'Solution': [OrderedDict([('number of solutions', 0), ('number of solutions displayed', 0)])]}

In [48]:
# print hasilnya
print(f"Optimal solusi: x1 = {model.x1()}, x2 = {model.x2()}, x3 = {model.x3()}, x4 = {model.x4()}")

Optimal solusi: x1 = 5.0, x2 = 0.0, x3 = 45.0, x4 = 0.0


In [50]:
print("Nilai optimumnya:", model.objective())

Nilai optimumnya: 1900.0


Di dalam melakukan perhitungan optimisasi yang bisa diaplikasikan berkali-kali dengan data yang berbeda-beda, sebaiknya data dan model dipisahkan. Untuk ini diperlukan model abstrak. Jika model di atas diubah menjadi model yang abstrak, maka bentuknya akan menjadi seperti berikut:  


$ objektif =  \sum_{x} untung_ix_i$  

Constraintnya adalah:
$ \sum_{x} JumlahPekerja_ix_i \le MaxJumlahPekerja$  
$ \sum_{x} BahanBakuA_ix_i \le MaxBahanBakuA$  
$ \sum_{x} BahanBakuB_ix_i \le MaxBahanBakuB$  
$ 𝑥_i \ge 0 $  


Dengan:  
i = nama produk, yakni produk 1, 2, 3, 4  
x = jumlah produk i yang diproduksi  

Contohnya adalah sebagai berikut:

In [107]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

In [108]:
# definisikan set
m.produk = Set()

# definisikan parameter
m.untung = Param(m.produk)
m.pekerja = Param(m.produk)
m.bahanbakuA = Param(m.produk)
m.bahanbakuB = Param(m.produk)
m.max_jumlah_pekerja = Param()
m.max_bahanbakuA = Param()
m.max_bahanbakuB = Param()

# definisikan variabel bebas
m.jumlah_produk = Var(m.produk, within=NonNegativeReals)

In [109]:
# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.untung[p]*m.jumlah_produk[p] for p in m.produk)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.keuntungan = Objective(rule=fungsi_objektif, sense=maximize)

In [110]:
# formulasikan constraint pekerja
def batas_pekerja(m):
    max_pekerja = sum(m.pekerja[p]*m.jumlah_produk[p] for p in m.produk) <= m.max_jumlah_pekerja
    return max_pekerja

# masukkan ke model
m.pekerja_cons = Constraint(rule=batas_pekerja)

In [111]:
# formulasikan constraint bahan baku A
def batas_A(m):
    max_A = sum(m.bahanbakuA[p]*m.jumlah_produk[p] for p in m.produk) <= m.max_bahanbakuA
    return max_A

# masukkan ke model
m.bahanbakuA_cons = Constraint(rule=batas_A)

In [112]:
# formulasikan constraint bahan baku B
def batas_B(m):
    max_B = sum(m.bahanbakuB[p]*m.jumlah_produk[p] for p in m.produk) <= m.max_bahanbakuB
    return max_B

# masukkan ke model
m.bahanbakuB_cons = Constraint(rule=batas_B)

In [115]:
# masukkan datanya
# parameter
data = {None: {
    'produk' : {None: [1, 2, 3, 4]},
    'untung' : {1: 20, 2: 12, 3: 40, 4: 25},
    'pekerja' : {1: 1, 2: 1, 3: 1, 4: 1},
    'bahanbakuA' : {1: 3, 2: 2, 3: 1, 4: 0},
    'bahanbakuB' : {1: 0, 2: 1, 3: 2, 4: 3},
    'max_jumlah_pekerja': {None: 50},
    'max_bahanbakuA': {None: 100},
    'max_bahanbakuB': {None: 90}
    }}

In [116]:
# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

Untuk memastikan apakah data tersebut telah masuk ke dalam model, kita bisa memanggil:  

    instance.pprint()

    
    

In [117]:
# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

In [119]:
# tulis dan tampilkan hasilnya
for produk in instance.jumlah_produk.values():
    print('Jumlah produksi', produk, 'sebanyak', produk.value)

print('Total keuntungan', instance.keuntungan())

Jumlah produksi jumlah_produk[1] sebanyak 5.0
Jumlah produksi jumlah_produk[2] sebanyak 0.0
Jumlah produksi jumlah_produk[3] sebanyak 45.0
Jumlah produksi jumlah_produk[4] sebanyak 0.0
Total keuntungan 1900.0


Berbeda dengan Concrete Model yang biasanya menggunakan nama "model" dan hasilnya pun dipanggil melalui "model.x" atau "model.hasil", Abstract Model menggunakan "instance" seperti contoh di atas.   
Praktik coding yang baik adalah dengan memeriksa fungsi objektif dan setiap constraint satu demi satu. Hal ini dilakukan dengan menonaktifkan constraint yang lain atau tidak mengetiknya sampai constraint sebelumnya benar. Pemeriksaan constraint ini dilakukan setelah memasukkan data dan menjalankan 

    instance.pprint()
    
    

<a id='04_03_01_network_optimization'></a>
#### 04.03.01 Network optimization (optimisasi jaringan supply)

Permasalahan ini merupakan salah satu topik dalam bidang rantai pasokan (supply chain). Terdapat beberapa lokasi pabrik, gudang, dan juga pelanggan. Masing-masing pabrik dan gudang memiliki kapasitas tertentu sementara pelanggan memiliki permintaan tertentu. Jarak dan ongkos transportasi antara satu lokasi ke lokasi lainnya diketahui. Pertanyaannya adalah dari mana sajakah (pabrik dan gudang) barang dikirim untuk memenuhi permintaan setiap pelanggan?

In [26]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.pabrik = Set()
m.gudang = Set()
m.pelanggan = Set()

# definisikan parameter
m.jarak_pabrik_gudang = Param(m.pabrik, m.gudang)
m.jarak_gudang_pelanggan = Param(m.gudang,m.pelanggan)
m.ongkos_pabrik_gudang = Param()
m.ongkos_gudang_pelanggan = Param()
m.max_barang_di_pabrik = Param(m.pabrik)
m.max_barang_di_gudang = Param(m.gudang)
m.permintaan = Param(m.pelanggan)

# definisikan variabel bebas
m.jumlah_barang_pabrik_gudang = Var(m.pabrik, m.gudang, bounds=(0,None))
m.jumlah_barang_gudang_pelanggan = Var(m.gudang, m.pelanggan, bounds=(0,None))

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.ongkos_pabrik_gudang*m.jarak_pabrik_gudang[pabrik,gudang]*m.jumlah_barang_pabrik_gudang[pabrik,gudang] for pabrik in m.pabrik for gudang in m.gudang)
    +sum(m.ongkos_gudang_pelanggan*m.jarak_gudang_pelanggan[gudang,pelanggan]*m.jumlah_barang_gudang_pelanggan[gudang,pelanggan] for gudang in m.gudang for pelanggan in m.pelanggan)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint supply dari pabrik
def max_supply_pabrik(m,pabrik):
    max_supply = sum(m.jumlah_barang_pabrik_gudang[pabrik,gudang] for gudang in m.gudang) <= m.max_barang_di_pabrik[pabrik]
    return max_supply

# masukkan ke model
m.supply_pabrik_cons = Constraint(m.pabrik, rule=max_supply_pabrik)

# formulasikan constraint kapasitas gudang
def max_supply_gudang(m,gudang):
    max_supply = sum(m.jumlah_barang_pabrik_gudang[pabrik,gudang] for pabrik in m.pabrik) <= m.max_barang_di_gudang[gudang]
    return max_supply

# masukkan ke model
m.supply_gudang_cons = Constraint(m.gudang, rule=max_supply_gudang)

# formulasikan constraint permintaan pelanggan
def permintaan_pelanggan(m,pelanggan):
    permintaan = sum(m.jumlah_barang_gudang_pelanggan[gudang,pelanggan] for gudang in m.gudang) >= m.permintaan[pelanggan]
    return permintaan

# masukkan ke model
m.permintaan_cons = Constraint(m.pelanggan, rule=permintaan_pelanggan)

# formulasikan constraint mass balance di gudang
def balance_gudang(m,gudang):
    balance = sum(m.jumlah_barang_pabrik_gudang[pabrik,gudang] for pabrik in m.pabrik) == sum(m.jumlah_barang_gudang_pelanggan[gudang,pelanggan] for pelanggan in m.pelanggan)
    return balance

# masukkan ke model
m.balance_cons = Constraint(m.gudang, rule=balance_gudang)
                                       
                                       
# masukkan datanya
# parameter
data = {None: {
    'pabrik' : {None: ['Cilegon', 'Tasikmalaya', 'Surakarta']},
    'gudang' : {None: ['Cikarang', 'Sidoarjo', 'Jepara']},
    'pelanggan' : {None: ['Jakarta', 'Bandung', 'Yogyakarta', 'Semarang', 'Surabaya', 'Banyuwangi', 'Cirebon', 'Malang']},
    'jarak_pabrik_gudang' : {('Cilegon','Cikarang'): 147, ('Cilegon','Sidoarjo'): 884, ('Cilegon','Jepara'): 613, 
                            ('Tasikmalaya','Cikarang'): 222, ('Tasikmalaya','Sidoarjo'): 664, ('Tasikmalaya','Jepara'): 393,
                            ('Surakarta','Cikarang'): 184, ('Surakarta','Sidoarjo'): 587, ('Surakarta','Jepara'): 317},
    'jarak_gudang_pelanggan' : {('Cikarang','Jakarta'): 42, ('Cikarang','Bandung'): 112, ('Cikarang','Yogyakarta'): 521, ('Cikarang','Semarang'): 403, ('Cikarang','Surabaya'): 741, ('Cikarang','Banyuwangi'): 1022, ('Cikarang','Cirebon'): 181, ('Cikarang','Malang'): 809,
                               ('Sidoarjo','Jakarta'): 782, ('Sidoarjo','Bandung'): 778, ('Sidoarjo','Yogyakarta'): 324, ('Sidoarjo','Semarang'): 350, ('Sidoarjo','Surabaya'): 31, ('Sidoarjo','Banyuwangi'): 284, ('Sidoarjo','Cirebon'): 572, ('Sidoarjo','Malang'): 71,
                               ('Jepara','Jakarta'): 518, ('Jepara','Bandung'): 514, ('Jepara','Yogyakarta'): 201, ('Jepara','Semarang'): 77, ('Jepara','Surabaya'): 427, ('Jepara','Banyuwangi'): 702, ('Jepara','Cirebon'): 309, ('Jepara','Malang'): 489},
    'ongkos_pabrik_gudang': {None: 0.08},
    'ongkos_gudang_pelanggan': {None: 0.1},
    'max_barang_di_pabrik': {'Cilegon': 1500, 'Tasikmalaya':1100, 'Surakarta':1100},
    'max_barang_di_gudang': {'Cikarang': 1000, 'Sidoarjo': 1300, 'Jepara': 1200},
    'permintaan': {'Jakarta':600, 'Bandung':450, 'Yogyakarta':500, 'Semarang':450, 'Surabaya':500, 'Banyuwangi':350, 'Cirebon':300, 'Malang':250}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

# tulis dan tampilkan hasilnya
for pabrik_gudang in instance.jumlah_barang_pabrik_gudang.values():
    print('Jumlah yang dikirim dari gudang', pabrik_gudang, 'sebanyak', pabrik_gudang.value)

for gudang_pelanggan in instance.jumlah_barang_gudang_pelanggan.values():
    print('Jumlah yang dikirim dari gudang', gudang_pelanggan, 'sebanyak', gudang_pelanggan.value)

print('Total ongkos', instance.ongkos())

                                       

Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Cilegon,Cikarang] sebanyak 1000.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Cilegon,Sidoarjo] sebanyak 0.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Cilegon,Jepara] sebanyak 200.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Tasikmalaya,Cikarang] sebanyak 0.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Tasikmalaya,Sidoarjo] sebanyak 100.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Tasikmalaya,Jepara] sebanyak 1000.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Surakarta,Cikarang] sebanyak 0.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Surakarta,Sidoarjo] sebanyak 1100.0
Jumlah yang dikirim dari gudang jumlah_barang_pabrik_gudang[Surakarta,Jepara] sebanyak 0.0
Jumlah yang dikirim dari gudang jumlah_barang_gudang_pelanggan[Cikarang,Jakarta] sebanyak 0.0
Jumlah yang dikirim dari gudang jumlah_barang_gudang_pelanggan

Model ini bisa digunakan untuk mengevaluasi berbagai macam skenario. Termasuk di antaranya adalah:  
- menentukan lokasi dan kapasitas pabrik
- menentukan lokasi dan kapasitas gudang
- mengevaluasi variasi permintaan
- mengevaluasi mode transportasi yang paling murah

Silakan diutak-atik angka dan data-datanya.

<a id='04_03_02_linear_blending'></a>
#### 04.03.02 Linear blending problem (proses pencampuran linier)
Jika kita punya berbagai produk dengan karakteristik tersendiri yang jika produk-produk tersebut dicampur maka karakteristik campuran bisa dihitung secara linier, maka formulasi di bawah ini bisa digunakan. 


In [1]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.produk = Set()

# definisikan parameter
m.biaya = Param(m.produk)
m.RON = Param(m.produk)
m.RON_min = Param()
m.jumlah_produk_min = Param()

# definisikan variabel bebas
m.jumlah_produk = Var(m.produk, bounds=(0,None))

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.biaya[produk]*m.jumlah_produk[produk] for produk in m.produk)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint RON minimum
def keperluan_RON(m):
    RON = sum(m.RON[produk]*m.jumlah_produk[produk] for produk in m.produk) >= m.RON_min*sum(m.jumlah_produk[produk] for produk in m.produk)
    return RON

# masukkan ke model
m.RON_cons = Constraint(rule=keperluan_RON)

# formulasikan constraint jumlah produk min
def keperluan_produk(m):
    jumlah = sum(m.jumlah_produk[produk] for produk in m.produk) >= m.jumlah_produk_min
    return jumlah

# masukkan ke model
m.keperluan_produk_cons = Constraint(rule=keperluan_produk)
                                     
# masukkan datanya
# parameter
data = {None: {
    'produk' : {None: ['RON80', 'RON90', 'RON95', 'RON100', 'RON110']},
    'biaya' : {'RON80': 80, 'RON90': 90, 'RON95': 95, 'RON100': 100, 'RON110': 110},
    'RON' : {'RON80': 80, 'RON90': 90, 'RON95': 95, 'RON100': 100, 'RON110': 110},
    'RON_min': {None: 98},
    'jumlah_produk_min': {None: 100}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

# tulis dan tampilkan hasilnya
for produk in instance.jumlah_produk.values():
    print('Jumlah produk', produk, 'adalah sebanyak', produk.value)

print('Total ongkos', instance.ongkos())

                                       

Jumlah produk jumlah_produk[RON80] adalah sebanyak 10.0
Jumlah produk jumlah_produk[RON90] adalah sebanyak 0.0
Jumlah produk jumlah_produk[RON95] adalah sebanyak 0.0
Jumlah produk jumlah_produk[RON100] adalah sebanyak 90.0
Jumlah produk jumlah_produk[RON110] adalah sebanyak 0.0
Total ongkos 9800.0


Silahkan diutak-atik angkanya dan diaplikasi ke masalah pencampuran linier lainnya.

<a id='04_03_03_assignment'></a>
#### 04.03.03 Assignment problem (masalah alokasi)
Di sini kita menentukan aktivitas apa yang akan dikerjakan di mesin yang mana. Permasalahan ini disebut juga sebagai permasalahan alokasi atau _assignment problem_.

In [5]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.aktivitas = Set()
m.mesin = Set()

# definisikan parameter
m.biaya = Param(m.aktivitas, m.mesin)

# definisikan variabel bebas
m.aktivitas_mesin = Var(m.aktivitas, m.mesin, bounds=(0,1)) # masalah ini diselesaikan dengan linear programming

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.biaya[aktivitas,mesin]*m.aktivitas_mesin[aktivitas,mesin] for aktivitas in m.aktivitas for mesin in m.mesin)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint satu aktivitas hanya menggunakan satu mesin
def satu(m,mesin):
    mesin = sum(m.aktivitas_mesin[aktivitas,mesin] for aktivitas in m.aktivitas) == 1
    return mesin

# masukkan ke model
m.mesin_cons = Constraint(m.mesin, rule=satu)

# formulasikan constraint satu mesin hanya boleh digunakan oleh satu aktivitas
def dua(m,aktivitas):
    aktivitas = sum(m.aktivitas_mesin[aktivitas,mesin] for mesin in m.mesin) == 1
    return aktivitas

# masukkan ke model
m.aktivitas_cons = Constraint(m.mesin, rule=dua)

# masukkan datanya
# parameter
data = {None: {
    'aktivitas' : {None: ['1', '2', '3', '4', '5']},
    'mesin' : {None: ['1', '2', '3', '4', '5']},
    'biaya' : {('1','1'): 5, ('1','2'): 6, ('1','3'): 10, ('1','4'): 8, ('1','5'): 6,
              ('2','1'): 4, ('2','2'): 6, ('2','3'): 4, ('2','4'): 5, ('2','5'): 3,
              ('3','1'): 4, ('3','2'): 4, ('3','3'): 3, ('3','4'): 4, ('3','5'): 9,
              ('4','1'): 6, ('4','2'): 8, ('4','3'): 6, ('4','4'): 2, ('4','5'): 10,
              ('5','1'): 3, ('5','2'): 9, ('5','3'): 8, ('5','4'): 7, ('5','5'): 6}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

# tulis dan tampilkan hasilnya
for aktivitas in instance.aktivitas_mesin.values():
    print(aktivitas, 'adalah', aktivitas.value)

print('Total ongkos', instance.ongkos())

                                       

aktivitas_mesin['1','1'] adalah 0.0
aktivitas_mesin['1','2'] adalah 1.0
aktivitas_mesin['1','3'] adalah 0.0
aktivitas_mesin['1','4'] adalah 0.0
aktivitas_mesin['1','5'] adalah 0.0
aktivitas_mesin['2','1'] adalah 0.0
aktivitas_mesin['2','2'] adalah 0.0
aktivitas_mesin['2','3'] adalah 0.0
aktivitas_mesin['2','4'] adalah 0.0
aktivitas_mesin['2','5'] adalah 1.0
aktivitas_mesin['3','1'] adalah 0.0
aktivitas_mesin['3','2'] adalah 0.0
aktivitas_mesin['3','3'] adalah 1.0
aktivitas_mesin['3','4'] adalah 0.0
aktivitas_mesin['3','5'] adalah 0.0
aktivitas_mesin['4','1'] adalah 0.0
aktivitas_mesin['4','2'] adalah 0.0
aktivitas_mesin['4','3'] adalah 0.0
aktivitas_mesin['4','4'] adalah 1.0
aktivitas_mesin['4','5'] adalah 0.0
aktivitas_mesin['5','1'] adalah 1.0
aktivitas_mesin['5','2'] adalah 0.0
aktivitas_mesin['5','3'] adalah 0.0
aktivitas_mesin['5','4'] adalah 0.0
aktivitas_mesin['5','5'] adalah 0.0
Total ongkos 17.0


<a id='04_03_04_project_selection'></a>
#### 04.03.04 Project selection (pemilihan proyek)
Pemilihan proyek didasarkan pada konsep *Discounted Return On Investment (DROI)*. DROI dihitung dengan membagi nilai *CAPEX* dengan nilai *NPV*. Oleh karena nilai *discounted rate* ada di dalam perhitungan *NPV*, maka nilai rasio antara *CAPEX* dan *NPV* ini disebut sebagai *DROI*.  

In [13]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.dana = Set()
m.proyek = Set()

# definisikan parameter
m.req_capex = Param(m.proyek)
m.npv = Param(m.proyek)
m.droi_proyek = Param(m.proyek)
m.droi_dana = Param(m.dana)
m.recom_capex = Param(m.proyek)
m.total_capex = Param(m.proyek)
m.max_budget = Param(m.dana)

# definisikan variabel bebas
m.aliran_dana = Var(m.dana, m.proyek, within=NonNegativeReals)
m.dana_takdigunakan = Var(m.dana, within=NonNegativeReals)
m.proyek_takdijalankan = Var(m.proyek, within=NonNegativeReals)

# formulasikan fungsi objektif adalah meminimumkan proyek yang tak dijalakan
def fungsi_objektif(m):
    obj = sum(m.proyek_takdijalankan[proyek] for proyek in m.proyek)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint dana
def satu(m, dana):
    expr = sum(m.aliran_dana[dana,proyek] for proyek in m.proyek) + m.dana_takdigunakan[dana] == m.max_budget[dana]
    return expr

# masukkan ke model
m.total_dana = Constraint(m.dana, rule=satu)

# formulasikan constraint projek
def dua(m,proyek):
    expr = sum(m.aliran_dana[dana,proyek] for dana in m.dana) + m.proyek_takdijalankan[proyek] == m.req_capex[proyek]
    return expr

# masukkan ke model
m.proyek_disupport = Constraint(m.proyek, rule=dua)

# formulasikan constraint limit dana
def empat(m,proyek):
    expr = sum(m.aliran_dana[dana,proyek] for dana in m.dana) <= m.recom_capex[proyek]
    return expr

# masukkan ke model
m.empat_cons = Constraint(m.proyek, rule=empat)

# formulasikan constraint profitability
def lima(m,proyek):
    expr = (m.droi_proyek[proyek]*m.req_capex[proyek])-(m.droi_proyek[proyek]*(m.req_capex[proyek]-sum(m.aliran_dana[dana,proyek] for dana in m.dana))) >= 0
    return expr

# masukkan ke model
m.lima_cons = Constraint(m.proyek, rule=lima)


# masukkan data parameter. 
data = {None: {
    'proyek' : {None: ['1', '2', '3']},
    'dana' : {None: ['1']},
    'max_budget': {'1':500.0},
    'req_capex' : {'1': 1.34, '2': 446.93, '3': 120.64},
    'total_capex' : {'1': 201.0, '2': 1674.49, '3': 125.14},
    'recom_capex' : {'1': 1.34, '2': 446.93, '3': 80.57},
    'npv' : {'1': 223.0, '2': 516.0, '3': 184.47},
    'droi_proyek' : {'1': 1.1, '2': 0.3, '3': 1.47}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)


In [14]:
# tulis dan tampilkan hasilnya
for aliran_dana in instance.aliran_dana.values():
    print(aliran_dana, 'adalah', aliran_dana.value)

print('Total ongkos', instance.ongkos())

aliran_dana['1','1'] adalah 1.34
aliran_dana['1','2'] adalah 446.93
aliran_dana['1','3'] adalah 51.73
Total ongkos 68.91


<a id='04_03_05_piecewise_linearization'></a>
#### 04.03.05 Piecewise Linearization

Piecewise linearization diambil dari contoh di link berikut ini:   
https://youtu.be/iPEDg5j2oyA


<a id='04_03_06_employee_scheduling'></a>
#### 04.03.06 Penjadwalan pekerja (employee scheduling)

Penjadwalan pekerja atau employee scheduling diambil dari contoh di link berikut ini:   
https://youtu.be/OjUmQRRfsEY


<a id='04_03_07_dynamic_programming'></a>
#### 04.03.07 Dynamic programming untuk secretary problem

Dynamic programming untuk secretary problem diambil dari contoh di link berikut ini:  
https://youtu.be/1oVW-V38yjo

<a id='04_03_08_kilang_minyak_bumi'></a>
#### 04.03.08 Kilang minyak bumi

Contoh-contoh lainnya akan ditambahkan nanti ke dalam notebook ini.

[Kembali ke daftar isi Linear Programming](#04_Linear_Programming)  
[Kembali ke daftar isi](#00_Daftar_Isi)

<a id='05_Integer_Programming'></a>
### 05. Integer programming
Integer Programming (IP) sering juga disebut sebagai masalah kombinasi (combinatorial problem). Untuk model-model yang melibatkan jumlah variabel yang tidak banyak, seharusnya model-model ini bisa diselesaikan dalam waktu yang cepat. Akan tetapi, untuk model yang melibatkan jumlah variabel yang sangat banyak (ribuan ke atas), maka model-model tersebut akan memakan waktu yang sangat lama untuk bisa diselesaikan. Dalam kasus ini, kita harus membuat heuristics untuk membantu agar solver dapat menyelesaikan permasalahan ini dalam waktu yang relatif singkat. Beberapa contoh sederhananya adalah:  

- [Alokasi mesin](#05_01_alokasi_mesin)
- [Lokasi pabrik](#05_02_plant_location)
- [Transisi energi](#05_03_energy_transition)
- [Penjadwalan proses batch](#05_04_process_scheduling)



<a id='05_01_alokasi_mesin'></a>
#### 05.01 Alokasi mesin

Masalah alokasi aktivitas-mesin di sub-bab LP di atas juga bisa diselesaikan dengan menggunakan integer.

In [7]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.aktivitas = Set()
m.mesin = Set()

# definisikan parameter
m.biaya = Param(m.aktivitas, m.mesin)

# definisikan variabel bebas
m.aktivitas_mesin = Var(m.aktivitas, m.mesin, within=Binary) # masalah ini diselesaikan dengan integer programming

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.biaya[aktivitas,mesin]*m.aktivitas_mesin[aktivitas,mesin] for aktivitas in m.aktivitas for mesin in m.mesin)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint satu aktivitas hanya menggunakan satu mesin
def satu(m,mesin):
    mesin = sum(m.aktivitas_mesin[aktivitas,mesin] for aktivitas in m.aktivitas) == 1
    return mesin

# masukkan ke model
m.mesin_cons = Constraint(m.mesin, rule=satu)

# formulasikan constraint satu mesin hanya boleh digunakan oleh satu aktivitas
def dua(m,aktivitas):
    aktivitas = sum(m.aktivitas_mesin[aktivitas,mesin] for mesin in m.mesin) == 1
    return aktivitas

# masukkan ke model
m.aktivitas_cons = Constraint(m.mesin, rule=dua)

# masukkan datanya
# parameter
data = {None: {
    'aktivitas' : {None: ['1', '2', '3', '4', '5']},
    'mesin' : {None: ['1', '2', '3', '4', '5']},
    'biaya' : {('1','1'): 5, ('1','2'): 6, ('1','3'): 10, ('1','4'): 8, ('1','5'): 6,
              ('2','1'): 4, ('2','2'): 6, ('2','3'): 4, ('2','4'): 5, ('2','5'): 3,
              ('3','1'): 4, ('3','2'): 4, ('3','3'): 3, ('3','4'): 4, ('3','5'): 9,
              ('4','1'): 6, ('4','2'): 8, ('4','3'): 6, ('4','4'): 2, ('4','5'): 10,
              ('5','1'): 3, ('5','2'): 9, ('5','3'): 8, ('5','4'): 7, ('5','5'): 6}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

# tulis dan tampilkan hasilnya
for aktivitas in instance.aktivitas_mesin.values():
    print(aktivitas, 'adalah', aktivitas.value)

print('Total ongkos', instance.ongkos())

                                       

aktivitas_mesin['1','1'] adalah 0.0
aktivitas_mesin['1','2'] adalah 1.0
aktivitas_mesin['1','3'] adalah 0.0
aktivitas_mesin['1','4'] adalah 0.0
aktivitas_mesin['1','5'] adalah 0.0
aktivitas_mesin['2','1'] adalah 0.0
aktivitas_mesin['2','2'] adalah 0.0
aktivitas_mesin['2','3'] adalah 0.0
aktivitas_mesin['2','4'] adalah 0.0
aktivitas_mesin['2','5'] adalah 1.0
aktivitas_mesin['3','1'] adalah 0.0
aktivitas_mesin['3','2'] adalah 0.0
aktivitas_mesin['3','3'] adalah 1.0
aktivitas_mesin['3','4'] adalah 0.0
aktivitas_mesin['3','5'] adalah 0.0
aktivitas_mesin['4','1'] adalah 0.0
aktivitas_mesin['4','2'] adalah 0.0
aktivitas_mesin['4','3'] adalah 0.0
aktivitas_mesin['4','4'] adalah 1.0
aktivitas_mesin['4','5'] adalah 0.0
aktivitas_mesin['5','1'] adalah 1.0
aktivitas_mesin['5','2'] adalah 0.0
aktivitas_mesin['5','3'] adalah 0.0
aktivitas_mesin['5','4'] adalah 0.0
aktivitas_mesin['5','5'] adalah 0.0
Total ongkos 17.0


<a id='05_02_plant_location'></a>
#### 05.02 Plant location (penetapan lokasi)
Biasanya ketika suatu pabrik ingin dibangun akan ada biaya tetap dan ketika pabrik tersebut beroperasi akan ada biaya operasional yang kontinyu. Contoh berikut menggabungkan formulasi biaya tetap (variabel binary) dan biaya kontinyu (variabel kontinyu).

In [9]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.pabrik = Set()
m.produk = Set()

# definisikan parameter
m.biaya = Param(m.pabrik, m.produk)
m.biaya_tetap = Param(m.pabrik)
m.permintaan = Param(m.produk)
m.kapasitas_pabrik = Param(m.pabrik)

# definisikan variabel bebas
m.jumlah_produk_pabrik = Var(m.pabrik, m.produk, bounds=(0,None)) # linear variabel
m.pabrik_ada = Var(m.pabrik, within=Binary) # integer variabel

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.biaya_tetap[pabrik]*m.pabrik_ada[pabrik] for pabrik in m.pabrik) 
    + sum(m.biaya[pabrik,produk]*m.jumlah_produk_pabrik[pabrik,produk] for pabrik in m.pabrik for produk in m.produk)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint setiap permintaan harus dipenuhi
def satu(m,produk):
    jumlahproduk = sum(m.jumlah_produk_pabrik[pabrik,produk] for pabrik in m.pabrik) >= m.permintaan[produk]
    return jumlahproduk

# masukkan ke model
m.mesin_cons = Constraint(m.produk, rule=satu)

# formulasikan constraint kapasitas pabrik
def dua(m,pabrik):
    kapasitas = sum(m.jumlah_produk_pabrik[pabrik,produk] for produk in m.produk) <= m.kapasitas_pabrik[pabrik]*m.pabrik_ada[pabrik]
    return kapasitas

# masukkan ke model
m.kapasitas_cons = Constraint(m.pabrik, rule=dua)

# masukkan datanya
# parameter
data = {None: {
    'pabrik' : {None: ['1', '2', '3', '4', '5']},
    'produk' : {None: ['1', '2', '3', '4', '5']},
    'biaya' : {('1','1'): 5, ('1','2'): 6, ('1','3'): 10, ('1','4'): 8, ('1','5'): 6,
              ('2','1'): 4, ('2','2'): 6, ('2','3'): 4, ('2','4'): 5, ('2','5'): 3,
              ('3','1'): 4, ('3','2'): 4, ('3','3'): 3, ('3','4'): 4, ('3','5'): 9,
              ('4','1'): 6, ('4','2'): 8, ('4','3'): 6, ('4','4'): 2, ('4','5'): 10,
              ('5','1'): 3, ('5','2'): 9, ('5','3'): 8, ('5','4'): 7, ('5','5'): 6},
    'biaya_tetap' : {'1': 50, '2': 60, '3': 100, '4': 80, '5': 60},
    'permintaan' : {'1': 500, '2': 400, '3': 300, '4': 580, '5': 460},
    'kapasitas_pabrik' : {'1': 500, '2': 400, '3': 500, '4': 400, '5': 600}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

# tulis dan tampilkan hasilnya
for produk in instance.jumlah_produk_pabrik.values():
    print(produk, 'adalah', produk.value)

print('Total ongkos', instance.ongkos())

                                       

jumlah_produk_pabrik['1','1'] adalah 40.0
jumlah_produk_pabrik['1','2'] adalah 0.0
jumlah_produk_pabrik['1','3'] adalah 300.0
jumlah_produk_pabrik['1','4'] adalah 0.0
jumlah_produk_pabrik['1','5'] adalah 0.0
jumlah_produk_pabrik['2','1'] adalah 0.0
jumlah_produk_pabrik['2','2'] adalah 0.0
jumlah_produk_pabrik['2','3'] adalah 0.0
jumlah_produk_pabrik['2','4'] adalah 80.0
jumlah_produk_pabrik['2','5'] adalah 320.0
jumlah_produk_pabrik['3','1'] adalah 0.0
jumlah_produk_pabrik['3','2'] adalah 0.0
jumlah_produk_pabrik['3','3'] adalah 0.0
jumlah_produk_pabrik['3','4'] adalah 500.0
jumlah_produk_pabrik['3','5'] adalah 0.0
jumlah_produk_pabrik['4','1'] adalah 260.0
jumlah_produk_pabrik['4','2'] adalah 0.0
jumlah_produk_pabrik['4','3'] adalah 0.0
jumlah_produk_pabrik['4','4'] adalah 0.0
jumlah_produk_pabrik['4','5'] adalah 140.0
jumlah_produk_pabrik['5','1'] adalah 200.0
jumlah_produk_pabrik['5','2'] adalah 400.0
jumlah_produk_pabrik['5','3'] adalah 0.0
jumlah_produk_pabrik['5','4'] adalah 0.0


<a id='05_03_energy_transition'></a>
#### 05.03 Transisi energi
Berikut ini formulasi sederhana dalam menentukan teknologi apa yang harus dibangun pada waktu kapan dan di mana. Keputusan yang akan diambil tentunya mempertimbangkan ketersediaan teknologi dan keperluan energi di setiap tempat yang dievaluasi. 

In [28]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel() 

# definisikan set
m.techs = Set()
m.products = Set()
m.suppliers = Set()
m.demands = Set()
m.years = Set()

# definisikan parameter untuk digunakan di dalam model
m.initial_demands = Param(m.demands,m.products)
m.growths = Param(m.demands,m.products)
m.yields = Param(m.techs,m.products,m.suppliers)
m.CO2emissions = Param(m.techs,m.suppliers)
m.TAC = Param(m.techs,m.suppliers)
m.max_installed = Param(m.techs,m.suppliers)

# definisikan variabel bebas
m.installed_techs = Var(m.techs, m.suppliers, m.years, within=Binary)
m.number_installed_techs= Var(m.techs, m.suppliers, m.years, bounds=(0,None))
m.totalTAC = Var(bounds=(None,None))
m.totalCO2 = Var(bounds=(None,None))
m.totalTACperyear = Var(m.years, bounds=(None, None))
m.totalCO2peryear = Var(m.years, bounds=(None, None))

# formulasikan fungsi objektif
def total_TAC(m):
    return sum(m.totalTACperyear[year] for year in m.years)
m.cost = Objective(rule=total_TAC, sense=minimize)

# fungsi-fungsi lain
def total_TAC_constraint(m):
    return m.totalTAC == sum(m.totalTACperyear[year] for year in m.years)

# masukkan ke model
m.totalTAC_cons = Constraint(rule=total_TAC_constraint)

def total_CO2(m):
    return m.totalCO2 == sum(m.totalCO2peryear[year] for year in m.years)

# masukkan ke model
m.totalCO2_cons = Constraint(rule=total_CO2)

def total_TACperyear(m,year):
    return m.totalTACperyear[year] == sum(m.number_installed_techs[tech,supplier,year]*m.TAC[tech,supplier] for tech in m.techs for supplier in m.suppliers)

# masukkan ke model
m.TACperyear_cons = Constraint(m.years, rule=total_TACperyear)

def total_CO2peryear(m,year):
    return m.totalCO2peryear[year] == sum(m.number_installed_techs[tech,supplier,year]*m.CO2emissions[tech,supplier] for tech in m.techs for supplier in m.suppliers) 

# masukkan ke model
m.CO2peryear_cons = Constraint(m.years, rule=total_CO2peryear)

# hitung parameter permintaan pertahun 
def demandperyear(m,demand,product,year):
    if year == 1:
        expression = m.initial_demands[demand,product]
        return expression
    expression = m.initial_demands[demand,product]*(1 + m.growths[demand,product]*year)
    return expression

# masukkan ke model
m.demandsperyear = Param(m.demands, m.products,m.years, initialize=demandperyear)

# formulasikan constraint setiap permintaan harus dipenuhi jika berada di daerah yang sama
def fulfill_demands(m,supplier,demand,product,year):
    if supplier == demand:
        expression = sum(m.number_installed_techs[tech,supplier,year]*m.yields[tech,product,supplier] for tech in m.techs) >= m.demandsperyear[demand,product,year]
        return expression
    return Constraint.Skip

# masukkan ke model
m.fulfill_demands_cons = Constraint(m.suppliers,m.demands,m.products,m.years, rule=fulfill_demands)

# formulasikan constraint kapasitas terinstall
def max_installed(m,tech,supplier):
    expression = sum(m.number_installed_techs[tech,supplier,year] for year in m.years) <= m.max_installed[tech,supplier]
    return expression

# masukkan ke model
m.max_installed_cons = Constraint(m.techs,m.suppliers, rule=max_installed)

# formulasikan constraint akumulasi kapasitas terinstall
def accum_installed(m,tech,supplier,year):
    if year == 1:
        expression = m.number_installed_techs[tech,supplier,year] == m.installed_techs[tech,supplier,year]
        return expression
    expression = m.number_installed_techs[tech,supplier,year] == m.number_installed_techs[tech,supplier,year-1] + m.installed_techs[tech,supplier,year]
    return expression

# masukkan ke model
m.accum_cons = Constraint(m.techs,m.suppliers,m.years, rule=accum_installed)

# masukkan datanya
# parameter
data = {None: {
    'techs' : {None: ['tek_A', 'tek_B', 'tek_C']},
    'products' : {None: ['prod_A', 'prod_B', 'prod_C']},
    'suppliers' : {None: ['daerah_A', 'daerah_B', 'daerah_C']},
    'demands' : {None: ['daerah_A', 'daerah_B', 'daerah_C']},
    'years' : {None: [1, 2, 3, 4, 5,6,7,8,9,10]},
    'initial_demands': {('daerah_A','prod_A'): 10.0, ('daerah_A','prod_B'): 10.0, ('daerah_A','prod_C'): 10.0,
                        ('daerah_B','prod_A'): 20.0, ('daerah_B','prod_B'): 20.0, ('daerah_B','prod_C'): 20.0,
                        ('daerah_C','prod_A'): 12.0, ('daerah_C','prod_B'): 12.0, ('daerah_C','prod_C'): 12.0},
    'growths': {('daerah_A','prod_A'): 0.02, ('daerah_A','prod_B'): 0.02, ('daerah_A','prod_C'): 0.02,
                        ('daerah_B','prod_A'): 0.02, ('daerah_B','prod_B'): 0.02, ('daerah_B','prod_C'): 0.02,
                        ('daerah_C','prod_A'): 0.02, ('daerah_C','prod_B'): 0.02, ('daerah_C','prod_C'): 0.02},
    'max_installed' : {('tek_A','daerah_A'): 100.0, ('tek_A','daerah_B'): 100.0, ('tek_A','daerah_C'): 100.0,
                        ('tek_B','daerah_A'): 100.0, ('tek_B','daerah_B'): 100.0, ('tek_B','daerah_C'): 100.0,
                        ('tek_C','daerah_A'): 100.0, ('tek_C','daerah_B'): 100.0, ('tek_C','daerah_C'): 100.0},
    'TAC' : {('tek_A','daerah_A'): 100.0, ('tek_A','daerah_B'): 100.0, ('tek_A','daerah_C'): 100.0,
                        ('tek_B','daerah_A'): 100.0, ('tek_B','daerah_B'): 100.0, ('tek_B','daerah_C'): 100.0,
                        ('tek_C','daerah_A'): 100.0, ('tek_C','daerah_B'): 100.0, ('tek_C','daerah_C'): 100.0},
    'CO2emissions' : {('tek_A','daerah_A'): 20.0, ('tek_A','daerah_B'): 20.0, ('tek_A','daerah_C'): 20.0,
                        ('tek_B','daerah_A'): 20.0, ('tek_B','daerah_B'): 20.0, ('tek_B','daerah_C'): 20.0,
                        ('tek_C','daerah_A'): 20.0, ('tek_C','daerah_B'): 20.0, ('tek_C','daerah_C'): 20.0},
    'yields': {('tek_A','prod_A','daerah_A'):8.0, ('tek_A','prod_A','daerah_B'):8.0, ('tek_A','prod_A','daerah_C'):8.0,
                    ('tek_A','prod_B','daerah_A'):8.0, ('tek_A','prod_B','daerah_B'):8.0, ('tek_A','prod_B','daerah_C'):8.0,
                    ('tek_A','prod_C','daerah_A'):8.0, ('tek_A','prod_C','daerah_B'):8.0, ('tek_A','prod_C','daerah_C'):8.0,
                    ('tek_B','prod_A','daerah_A'):8.0, ('tek_B','prod_A','daerah_B'):8.0, ('tek_B','prod_A','daerah_C'):8.0,
                    ('tek_B','prod_B','daerah_A'):8.0, ('tek_B','prod_B','daerah_B'):8.0, ('tek_B','prod_B','daerah_C'):8.0,
                    ('tek_B','prod_C','daerah_A'):8.0, ('tek_B','prod_C','daerah_B'):8.0, ('tek_B','prod_C','daerah_C'):8.0,
                    ('tek_C','prod_A','daerah_A'):8.0, ('tek_C','prod_A','daerah_B'):8.0, ('tek_C','prod_A','daerah_C'):8.0,
                    ('tek_C','prod_B','daerah_A'):8.0, ('tek_C','prod_B','daerah_B'):8.0, ('tek_C','prod_B','daerah_C'):8.0,
                    ('tek_C','prod_C','daerah_A'):8.0, ('tek_C','prod_C','daerah_B'):8.0, ('tek_C','prod_C','daerah_C'):8.0}
   
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)
# instance.pprint()

# # jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

In [29]:
# tulis dan tampilkan hasilnya
for tech in instance.installed_techs.values():
    print(tech.value)

1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


In [30]:
for number_of_techs_installed in instance.number_installed_techs.values():
    print(number_of_techs_installed.value)

1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0


In [31]:
for TACperyear in instance.totalTACperyear.values():
    print(TACperyear.value)

700.0
700.0
700.0
700.0
700.0
700.0
700.0
700.0
700.0
700.0


In [32]:
for CO2peryear in instance.totalCO2peryear.values():
    print(CO2peryear.value)

140.0
140.0
140.0
140.0
140.0
140.0
140.0
140.0
140.0
140.0


In [33]:
print('Total TAC', instance.totalTAC())
print('Total CO2', instance.totalCO2())

Total TAC 7000.0
Total CO2 1400.0


<a id='05_04_process_scheduling'></a>
#### 05.04 Process scheduling (penjadwalan proses batch)

Penjadwalan proses batch dapat diformulasikan secara sederhana menjadi:  
$ x_{ijt} = 1$ jika proses i dilakukan dengan reaktor j, dan 0 jika sebaliknya  
$ \sum_j(x_{ijt}) = 1$, masing-masing proses HARUS dilakukan di dalam satu reaktor  
$ \sum_i(x_{ijt}) \le 1$, masing-masing reaktor hanya boleh digunakan untuk maksimum satu proses  
$ \sum_j(x_{ijt}) \le 1$, masing-masing proses MAX dilakukan di dalam satu reaktor  
$ x_{ijt}demand_i - kapasitas_j \le 0$, hanya reaktor dengan kapasitas yang cukup yang digunakan untuk memproduksi suatu produk dengan kapasitas tertentu  

Fungsi objektifnya adalah meminimumkan:  
$ \sum_{ijt}(x_{ijt}kapasitas_j/demand_i) $

i = produk  
j = proses atau reaktor  
t = timeslot  

In [10]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.proses = Set()
m.produk = Set()
m.timeslot = Set()

# definisikan parameter
m.kapasitas_reaktor = Param(m.proses)
m.demand = Param(m.produk)

# definisikan variabel bebas
m.jadwal = Var(m.produk, m.proses, m.timeslot, within=Binary)

# formulasikan fungsi objektif adalah meminimumkan proyek yang tak dijalakan
def fungsi_objektif(m):
    obj = sum(m.jadwal[produk,proses,timeslot]*m.kapasitas_reaktor[proses]/m.demand[produk] for produk in m.produk for proses in m.proses for timeslot in m.timeslot)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint masing-masing produk HARUS dibuat di satu reaktor di satu waktu
def satu(m,produk):
    expr = sum(m.jadwal[produk,proses,timeslot] for proses in m.proses for timeslot in m.timeslot) == 1
    return expr

# masukkan ke model
m.satu_cons = Constraint(m.produk, rule=satu)

# formulasikan constraint masing-masing reaktor dalam satu waktu cuma boleh dipake sekali
def dua(m,proses, timeslot):
    expr = sum(m.jadwal[produk,proses,timeslot] for produk in m.produk) <= 1
    return expr

# masukkan ke model
m.dua_cons = Constraint(m.proses, m.timeslot, rule=dua)

# formulasikan constraint masing-masing produk MAKS dibuat di satu reaktor di satu waktu
def tiga(m,produk):
    expr = sum(m.jadwal[produk,proses,timeslot] for proses in m.proses for timeslot in m.timeslot) <= 1
    return expr

# masukkan ke model
m.tiga_cons = Constraint(m.produk, rule=tiga)

# formulasikan limit kapasitas reaktor
def empat(m,produk, proses, timeslot):
    expr = m.demand[produk]*m.jadwal[produk,proses,timeslot]-m.kapasitas_reaktor[proses] <= 0
    return expr

# masukkan ke model
m.empat_cons = Constraint(m.produk, m.proses, m.timeslot, rule=empat)


# masukkan data parameter. 
data = {None: {
    'proses' : {None: ['reaktor_1', 'reaktor_2', 'reaktor_3', 'reaktor_4']},
    'produk' : {None: ['Lily-white', 'Cool-yellow', 'Sunset-orange', 
                      'Off-white', 'Sky-blue', 'Flame-red', 'Coal-black',
                      'Blush-pink','Bottle-green', 'Navy-blue', 'Lime-green',
                      'Turqoise']},
    'timeslot': {None: ['day-1-time-1', 'day-1-time-2', 'day-1-time-3', 'day-1-time-4',
                'day-2-time-1', 'day-2-time-2', 'day-2-time-3', 'day-2-time-4',
                'day-3-time-1', 'day-3-time-2', 'day-3-time-3', 'day-3-time-4',
                'day-4-time-1', 'day-4-time-2', 'day-4-time-3', 'day-4-time-4']},
    'kapasitas_reaktor' : {'reaktor_1': 100.0, 'reaktor_2': 300.0, 
                           'reaktor_3': 100.0, 'reaktor_4': 400.0},
    'demand' : {'Lily-white': 200.0, 'Cool-yellow': 100.0, 'Sunset-orange':300.0, 
                      'Off-white':400.0, 'Sky-blue':200.0, 'Flame-red':100.0, 'Coal-black':300.0,
                      'Blush-pink':400.0,'Bottle-green':200.0, 'Navy-blue':100.0, 'Lime-green':300.0,
                      'Turqoise':400.0}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

In [11]:
# tulis dan tampilkan hasilnya
for jadwal in instance.jadwal.values():
    print(jadwal, 'adalah', jadwal.value)

print('Total ongkos', instance.ongkos())

jadwal[Lily-white,reaktor_1,day-1-time-1] adalah 0.0
jadwal[Lily-white,reaktor_1,day-1-time-2] adalah 0.0
jadwal[Lily-white,reaktor_1,day-1-time-3] adalah 0.0
jadwal[Lily-white,reaktor_1,day-1-time-4] adalah 0.0
jadwal[Lily-white,reaktor_1,day-2-time-1] adalah 0.0
jadwal[Lily-white,reaktor_1,day-2-time-2] adalah 0.0
jadwal[Lily-white,reaktor_1,day-2-time-3] adalah 0.0
jadwal[Lily-white,reaktor_1,day-2-time-4] adalah 0.0
jadwal[Lily-white,reaktor_1,day-3-time-1] adalah 0.0
jadwal[Lily-white,reaktor_1,day-3-time-2] adalah 0.0
jadwal[Lily-white,reaktor_1,day-3-time-3] adalah 0.0
jadwal[Lily-white,reaktor_1,day-3-time-4] adalah 0.0
jadwal[Lily-white,reaktor_1,day-4-time-1] adalah 0.0
jadwal[Lily-white,reaktor_1,day-4-time-2] adalah 0.0
jadwal[Lily-white,reaktor_1,day-4-time-3] adalah 0.0
jadwal[Lily-white,reaktor_1,day-4-time-4] adalah 0.0
jadwal[Lily-white,reaktor_2,day-1-time-1] adalah 0.0
jadwal[Lily-white,reaktor_2,day-1-time-2] adalah 1.0
jadwal[Lily-white,reaktor_2,day-1-time-3] adal

jadwal[Off-white,reaktor_4,day-2-time-4] adalah 0.0
jadwal[Off-white,reaktor_4,day-3-time-1] adalah 0.0
jadwal[Off-white,reaktor_4,day-3-time-2] adalah 0.0
jadwal[Off-white,reaktor_4,day-3-time-3] adalah 0.0
jadwal[Off-white,reaktor_4,day-3-time-4] adalah 0.0
jadwal[Off-white,reaktor_4,day-4-time-1] adalah 0.0
jadwal[Off-white,reaktor_4,day-4-time-2] adalah 0.0
jadwal[Off-white,reaktor_4,day-4-time-3] adalah 0.0
jadwal[Off-white,reaktor_4,day-4-time-4] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-1-time-1] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-1-time-2] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-1-time-3] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-1-time-4] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-2-time-1] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-2-time-2] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-2-time-3] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-2-time-4] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-3-time-1] adalah 0.0
jadwal[Sky-blue,reaktor_1,day-3-time-2] adalah 0.0
jadwal[Sky-blue,reakto

jadwal[Bottle-green,reaktor_2,day-1-time-4] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-2-time-1] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-2-time-2] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-2-time-3] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-2-time-4] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-3-time-1] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-3-time-2] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-3-time-3] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-3-time-4] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-4-time-1] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-4-time-2] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-4-time-3] adalah 0.0
jadwal[Bottle-green,reaktor_2,day-4-time-4] adalah 1.0
jadwal[Bottle-green,reaktor_3,day-1-time-1] adalah 0.0
jadwal[Bottle-green,reaktor_3,day-1-time-2] adalah 0.0
jadwal[Bottle-green,reaktor_3,day-1-time-3] adalah 0.0
jadwal[Bottle-green,reaktor_3,day-1-time-4] adalah 0.0
jadwal[Bottle-green,reaktor_3,day-2-time-1] adalah 0.0
jadwal[Bot

jadwal[Turqoise,reaktor_4,day-4-time-2] adalah 0.0
jadwal[Turqoise,reaktor_4,day-4-time-3] adalah 0.0
jadwal[Turqoise,reaktor_4,day-4-time-4] adalah 1.0
Total ongkos 13.5


Demikianlah beberapa contoh Integer dan Linear Programming yang dapat disampaikan. Kasus-kasus lainnya dapat ditambahkan di kemudian hari. 

[Kembali ke daftar isi Integer Programming](#05_Integer_Programming)  
[Kembali ke daftar isi](#00_Daftar_Isi)

<a id='06_Nonlinear_Programming'></a>
### 06. Nonlinear programming
Seperti namanya, NonLinear Programming (NLP) memiliki formula yang tidak linier, baik di fungsi obyektifnya dan/atau di constraintnya. Di bab [Constrained Optimization](#07_02_Constrained_Optimization) sebelumnya kita telah menggunakan paket _scipy.optimize_ dan metode _stochastic algorithm_ untuk menyelesaikan permasalahan NLP. Di bagian ini kita akan coba gunakan _pyomo_ hanya untuk melengkapi contoh-contoh yang telah ada. 

Mencari nilai jari-jari dan tinggi suatu silinder dengan volume tertentu agar luas areanya minimum.  
$Area = 2\pi{}r(r+h) $  
$Volume = \pi{}hr^2$

In [16]:
from pyomo.environ import *
m = ConcreteModel()
m.r = Var(bounds=(0,None))
m.h = Var(bounds=(0,None))

In [17]:
import math
def luas_area(m):
    return 2*math.pi*m.r*(m.r+m.h)
m.luas_area_obj = Objective(rule=luas_area, sense=minimize)

In [18]:
m.vol =Param(initialize=400)
def volume(m):
    return math.pi*m.h*m.r**2 == m.vol
m.volume_con = Constraint(rule=volume)

Di sini, solver yang akan digunakan adalah **ipopt** dan lokasi executable filenya harus didefinisikan seperti contoh di bawah. Gunakan "/" sebagai pemisah antara folder.  
Jika code ini dijalankan dengan _spyder_, maka lokasi executable file **ipopt** ini tidak perlu didefinisikan.

In [19]:
solver = SolverFactory('ipopt', executable="D:/Dropbox/Proyek buku Optimisasi di Teknik Kimia/Pyomo/Ipopt-3.11.1-win64-intel13.1/bin/ipopt.exe")
solver.solve(m)

{'Problem': [{'Lower bound': -inf, 'Upper bound': inf, 'Number of objectives': 1, 'Number of constraints': 1, 'Number of variables': 2, 'Sense': 'unknown'}], 'Solver': [{'Status': 'ok', 'Message': 'Ipopt 3.11.1\\x3a Optimal Solution Found', 'Termination condition': 'optimal', 'Id': 0, 'Error rc': 0, 'Time': 0.12198233604431152}], 'Solution': [OrderedDict([('number of solutions', 0), ('number of solutions displayed', 0)])]}

In [25]:
print('Jari-jari = {0:.1f}'.format(m.r()))
print()
print('Tinggi = {0:.1f}'.format(m.h()))
print()
print('Luas Area = {0:.0f}'.format(m.luas_area_obj()))
print()
print('Volume = {0:.0f}'.format(m.volume_con()))


Jari-jari = 4.0

Tinggi = 8.0

Luas Area = 301

Volume = 400


Silahkan mengganti nilai volumenya atau juga menggunakan rumus-rumus atau permasalahan nonlinear lainnya. 

Contoh lainnya adalah meminimumkan fungsi Rosenbrock

In [21]:
from pyomo.environ import *
model = ConcreteModel()
model.x = Var(initialize=1.5)
model.y = Var(initialize=1.5)
def rosenbrock(m):
    return (1.0-m.x)**2 + 100.0*(m.y - m.x**2)**2
model.obj = Objective(rule=rosenbrock, sense=minimize)
SolverFactory('ipopt', executable="D:/Dropbox/Proyek buku Optimisasi di Teknik Kimia/Pyomo/Ipopt-3.11.1-win64-intel13.1/bin/ipopt.exe").solve(model, tee=True)
model.pprint()

Ipopt 3.11.1: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt
******************************************************************************

NOTE: You are using Ipopt by default with the MUMPS linear solver.
      Other linear solvers might be more efficient (see Ipopt documentation).


This is Ipopt version 3.11.1, running with linear solver mumps.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        3

Total number of variables............................:        2
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0


In [23]:
# kita print hasilnya
print(f"Nilai x = {model.x()}, nilai y = {model.y()}")
print()
print(f"Nilai minimum fungsi obyektifnya = {model.obj()}")
print()

Nilai x = 1.0000000000008233, nilai y = 1.0000000000016314

Nilai minimum fungsi obyektifnya = 7.013645951336496e-25



Contoh-contoh lainnya akan ditambahkan seiring dengan berjalannya waktu, pengalaman, dan studi kasus nyata :)

[Kembali ke daftar isi](#00_Daftar_Isi)

<a id='07_Stochastic_Programming'></a>
### 07. Stochastic programming

Stochastic programming berhubungan dengan permasalahan yang memiliki parameter yang tidak konstan dan distribusi kemungkinannya (probability distribution) dapat diketahui atau diprediksi. Tujuan utama dari stochastic programming adalah untuk mendapatkan keputusan rata-rata yang optimum (average optimum decision) dengan parameter yang tidak konstan tersebut. Salah satu pendekatan penyelesaiannya adalah dengan 2-stage stochastic programming. 
- Stage 1 adalah mencari variabel keputusannya (_decision variable_) yang secara rata-rata untuk semua skenario dapat memberikan nilai maksimum atau minimum di fungsi objektifnya, 
- Stage 2 adalah mencari nilai maksimum atau minimum fungsi objektif tersebut berdasarkan keputusan yang telah diambil di Stage 1 sebelumnya berdasarkan kemungkinan terjadinya skenario-skenario yang kita mungkin hadapi. 

Contoh pertama yang akan kita diskusikan diambil dari link berikut: https://www.youtube.com/watch?v=UZCuyeLRvsw  
  
Kita akan membooking Hotel Wiltmore untuk konferensi selama 3 hari. Datanya:  
- Maksimum kamar yang bisa dibooking adalah 600  
- Klo pesan sekarang, harganya didiskon menjadi 160 per kamar
- Harga penuhnya 212 per kamar
- Klo terlebih booking, atau yang datang kurang, kita harus membayar penuh. Berarti kita akan menambah 212-160 = 52 per kamar
- Klo kurang booking, atau yang datang kebanyakan, kita harus book hotel terdekat (170 per kamar) dan transportasinya (80 per kamar). Total tambahan adalah 250 per kamar.

Berikut ini adalah data kemungkinan 6 skenario yang mungkin terjadi:  

|Skenario|Kemungkinan|Yang datang|
|---|---|---|
|1|10%|300|
|2|20%|400|
|3|25%|500|
|4|20%|600|
|5|15%|700|
|6|10%|800|


Stage 1: variabelnya adalah jumlah kamar yang akan dibooking (r)  
Stage 2: variabelnya adalah jumlah kamar yang terlebih (o) dan terkurang booking (u)

Model matematikanya adalah:  
minimumkan $3[160r +10\%(52o_1+250u_1)+20\%(52o_2+250u_2)+25\%(52o_3+250u_3)+20\%(52o_4+250u_4)+15\%(52o_5+250u_5)+10\%(52o_6+250u_6)]$  

Constraint di Stage 1:  
$r \le 600 $ --> jumlah kamar yang bisa dibooking   

Constraint di Stage 2: jumlah kamar yang telah dibooking - terlebih booking + terkurang booking = yang datang di setiap kemungkinan.  

$r - o_1 + u_1 = 300$  
$r - o_2 + u_2 = 400$  
$r - o_3 + u_3 = 500$  
$r - o_4 + u_4 = 600$  
$r - o_5 + u_5 = 700$  
$r - o_6 + u_6 = 80$  

$r, o, u \ge 0 $  
Subscript 1 s/d 6 adalah skenario. 

In [3]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.skenario = Set()

# definisikan parameter
m.biaya_diskon = Param(initialize=160)
m.max_kamar = Param(initialize=600)
m.biaya_overbooked = Param(initialize=52)
m.biaya_underbooked = Param(initialize=250)
m.jumlah_hari = Param(initialize=3)
m.yang_datang = Param(m.skenario)
m.kemungkinan = Param(m.skenario)

# definisikan variabel bebas
m.r = Var(within=NonNegativeReals)
m.o = Var(m.skenario, within=NonNegativeReals)
m.u = Var(m.skenario, within=NonNegativeReals)

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = m.jumlah_hari*(m.biaya_diskon*m.r+sum(m.kemungkinan[skenario]*(m.biaya_overbooked*m.o[skenario] + m.biaya_underbooked*m.u[skenario]) for skenario in m.skenario))
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=minimize)

# formulasikan constraint 
def satu(m):
    expr = m.r <= m.max_kamar
    return expr
m.satu_cons = Constraint(rule=satu)

def dua(m,skenario):
    expr = m.r - m.o[skenario] + m.u[skenario] == m.yang_datang[skenario]
    return expr
m.dua_cons = Constraint(m.skenario, rule=dua)

# masukkan datanya
data = {None: {
    'biaya_diskon' : {None: 160},
    'biaya_overbooked' : {None: 52},
    'biaya_underbooked' : {None: 250},
    'jumlah_hari' : {None: 3},
    'max_kamar' : {None: 600},
    'skenario' : {None: ['1', '2', '3', '4', '5', '6']},
    'yang_datang' : {'1': 300.0, '2': 400.0, '3':500.0, '4':600.0, '5':700.0, '6':800.0},
    'kemungkinan' : {'1': 0.1, '2': 0.2, '3':0.25, '4':0.2, '5':0.15, '6':0.1}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)



                                       

In [4]:
# tulis dan tampilkan hasilnya
print('Kamar yang dibook', instance.r())
print()
print('Total biaya', instance.ongkos())
print()
for skenario in instance.o.values():
    print('Kamar yang ovebook di skenario ', skenario, 'adalah', skenario.value)
print()
for skenario in instance.u.values():
    print('Kamar yang ovebook di skenario ', skenario, 'adalah', skenario.value)


Kamar yang dibook 400.0

Total biaya 306060.0

Kamar yang ovebook di skenario  o['1'] adalah 100.0
Kamar yang ovebook di skenario  o['2'] adalah 0.0
Kamar yang ovebook di skenario  o['3'] adalah 0.0
Kamar yang ovebook di skenario  o['4'] adalah 0.0
Kamar yang ovebook di skenario  o['5'] adalah 0.0
Kamar yang ovebook di skenario  o['6'] adalah 0.0

Kamar yang ovebook di skenario  u['1'] adalah 0.0
Kamar yang ovebook di skenario  u['2'] adalah 0.0
Kamar yang ovebook di skenario  u['3'] adalah 100.0
Kamar yang ovebook di skenario  u['4'] adalah 200.0
Kamar yang ovebook di skenario  u['5'] adalah 300.0
Kamar yang ovebook di skenario  u['6'] adalah 400.0


Hasil ini bisa dibaca seperti "untuk semua skenario yang telah kita identifikasi ini, secara rata-rata kita akan mengalami biaya terendah jika membook kamar sebanyak 400 kamar dan biayanya adalah 306060.  

Tentunya data ini bisa diubah-ubah untuk melihat bagaimana hasilnya akan bereaksi. Dari sana, kita bisa menentukan sikap kita sendiri. 

Contoh kedua adalah mengenai jumlah tempat duduk untuk setiap kelas di pesawat terbang. Contoh ini diambil dari link berikut https://www.youtube.com/watch?v=VzI9U7yeVT0  

Kita akan membeli pesawat yang memiliki sejumlah tempat duduk First Class, Business Class, dan Economy Class. Datanya:  
- Maksimum space yang tersedia adalah 190,  
- Space untuk kelas First, Business, dan Economy adalah 2, 1.5, dan 1, secara berurutan,
- Harga kelas First, Business, dan Economy adalah 1600, 1100, dan 500, secara berurutan.

Berikut ini adalah data kemungkinan 3 skenario yang mungkin terjadi:  

|Skenario|Kemungkinan|Permintaan First| Permintaan Business|Permintaan Economy|
|---|---|---|---|---|
|1|40%|25|60|210|
|2|30%|12|30|170|
|3|30%|5|9|150|


Stage 1: variabelnya adalah jumlah seat untuk setiap kelas (F,B,E) yang akan dibeli  
Stage 2: variabelnya adalah jumlah seat yang terjual untuk setiap skenario di setiap kelas  

Model matematikanya adalah:  
**maximumkan**  
$40\%(1600SF_1 + 1100SB_1 + 500SE_1)+30\%(1600SF_1 + 1100SB_1 + 500SE_1)+30\%(1600SF_1 + 1100SB_1 + 500SE_1)$  

Constraint di Stage 1:  
$2F + 1.5B + E \le 190 $ --> jumlah space yang tersedia   

Constraint di Stage 2: jumlah seat yang terjual di setiap skenario.  

Skenario 1:  
$SF_1 \le F$  
$SB_1 \le B$  
$SE_1 \le E$  
$SF_1 \le 25$  
$SB_1 \le 60$  
$SE_1 \le 210$  

Skenario 2:  
$SF_2 \le F$  
$SB_2 \le B$  
$SE_2 \le E$  
$SF_2 \le 12$  
$SB_2 \le 30$  
$SE_2 \le 170$  

Skenario 3:  
$SF_3 \le F$  
$SB_3 \le B$  
$SE_3 \le E$  
$SF_3 \le 5$  
$SB_3 \le 9$  
$SE_3 \le 150$  

$F, B, E, SF, SB, SE \ge 0 $  
Subscript 1 s/d 3 adalah skenario. 


In [12]:
from pyomo.environ import *

# definisikan modelnya
m = AbstractModel()

# definisikan set
m.kelas = Set()
m.skenario = Set()

# definisikan parameter
m.space = Param(m.kelas)
m.max_space = Param(initialize = 190)
m.revenue = Param(m.kelas)
m.permintaan = Param(m.kelas, m.skenario)
m.kemungkinan = Param(m.skenario)

# definisikan variabel bebas
m.seat_dibeli = Var(m.kelas, within=NonNegativeReals)
m.seat_terjual = Var(m.kelas, m.skenario, within=NonNegativeReals)

# formulasikan fungsi objektif
def fungsi_objektif(m):
    obj = sum(m.kemungkinan[skenario]*sum(m.revenue[kelas]*m.seat_terjual[kelas,skenario] for kelas in m.kelas) for skenario in m.skenario)
    return obj

# definisikan fungsi objektif ke dalam Abstract Model
m.ongkos = Objective(rule=fungsi_objektif, sense=maximize)

# formulasikan constraint 
def satu(m, kelas, skenario):
    expr = m.seat_terjual[kelas,skenario] <= m.seat_dibeli[kelas]
    return expr
m.satu_cons = Constraint(m.kelas, m.skenario, rule=satu)

def dua(m,kelas, skenario):
    expr = m.seat_terjual[kelas,skenario] <= m.permintaan[kelas, skenario]
    return expr
m.dua_cons = Constraint(m.kelas, m.skenario, rule=dua)

def tiga(m):
    expr = sum(m.seat_dibeli[kelas]*m.space[kelas] for kelas in m.kelas) <= m.max_space
    return expr
m.tiga_cons = Constraint(rule=tiga)

# masukkan datanya
data = {None: {
    'max_space' : {None: 190},
    'kelas' : {None: ['First', 'Business', 'Economy']},
    'skenario' : {None: ['1', '2', '3']},
    'revenue' : {'First': 1600.0, 'Business': 1100.0, 'Economy':500.0},
    'space' : {'First': 2.0, 'Business': 1.5, 'Economy':1.0},
    'permintaan' : {('First', '1'): 25.0, ('Business','1'): 60.0, ('Economy','1'):210.0,
                   ('First', '2'): 12.0, ('Business','2'): 30.0, ('Economy','2'):170.0,
                   ('First', '3'): 5.0, ('Business','3'): 9.0, ('Economy','3'):150.0},
    'kemungkinan' : {'1': 0.4, '2': 0.3, '3':0.3}
    }}

# buat instance atau memasukkan data di atas ke dalam Abstract Model
instance = m.create_instance(data)

# jalankan optimisasinya
solver = SolverFactory('glpk')
hasil = solver.solve(instance)

In [17]:
# tulis dan tampilkan hasilnya
for seat in instance.seat_dibeli.values():
    print('Seat', seat,'yang dibeli adalah', seat.value)
print()
print('Total biaya', instance.ongkos())
print()
for seat in instance.seat_terjual.values():
    print('Seat yang terjual di skenario ', seat, 'adalah', seat.value)
print()

Seat seat_dibeli[First] yang dibeli adalah 12.0
Seat seat_dibeli[Business] yang dibeli adalah 30.0
Seat seat_dibeli[Economy] yang dibeli adalah 121.0

Total biaya 102410.0

Seat yang terjual di skenario  seat_terjual[First,'1'] adalah 12.0
Seat yang terjual di skenario  seat_terjual[First,'2'] adalah 12.0
Seat yang terjual di skenario  seat_terjual[First,'3'] adalah 5.0
Seat yang terjual di skenario  seat_terjual[Business,'1'] adalah 30.0
Seat yang terjual di skenario  seat_terjual[Business,'2'] adalah 30.0
Seat yang terjual di skenario  seat_terjual[Business,'3'] adalah 9.0
Seat yang terjual di skenario  seat_terjual[Economy,'1'] adalah 121.0
Seat yang terjual di skenario  seat_terjual[Economy,'2'] adalah 121.0
Seat yang terjual di skenario  seat_terjual[Economy,'3'] adalah 121.0



Seperti contoh sebelumnya, untuk data dan kemungkinan skenario yang ada sekarang, seat optimum yang harus dibeli untuk kelas First, Business, dan Economy adalah 12, 30, dan 121, secara berurutan. Nilai revenue rata-rata untuk skenario-skenario ini adalah 102410. 

[Kembali ke daftar isi](#00_Daftar_Isi)

<a id='08_Robust_Programming'></a>
### 08. Robust programming

Robust programming secara singkat adalah melakukan optimisasi di dalam kondisi terburuk. Setiap parameter yang tidak pasti akan memiliki nilai minimum dan maksimum. Dengan ini, formulasi yang akan diselesaikan di robust programming menggunakan nilai minimum atau maksimum untuk setiap parameter yang tidak pasti tersebut tergantung dari kondisinya. 

Misalnya untuk fungsi objektif seperti berikut:  
$ \min  c_1x_1 + c_2x_2 $  
Kita harus menggunakan nilai $c_1$ dan $c_2$ yang paling **besar**.  

Sementara untuk fungsi objektif seperti:  
$ \max  c_1x_1 + c_2x_2 $  
Kita harus menggunakan nilai $c_1$ dan $c_2$ yang paling **kecil**.  

Demikian halnya dengan constraint, yakni:  
$ a_1x_1 + a_2x_2 \le b$  
Maka, secara worst case, kita harus memilih nilai **terbesar** untuk $a_1$ dan $a_2$ dan nilai **terkecil** untuk $b_2$.  

Sebaliknya:
$ a_1x_1 + a_2x_2 \ge b$  
Maka, secara worst case, kita harus memilih nilai **terkecil** untuk $a_1$ dan $a_2$ dan nilai **terbesar** untuk $b_2$.  

Di mana:  
$ c_1  \epsilon  [c_1^{min}; c_1^{max}]$  
$ c_2  \epsilon  [c_2^{min}; c_2^{max}]$  
$ a_1  \epsilon  [a_1^{min}; a_1^{max}]$  
$ a_2  \epsilon  [a_2^{min}; a_2^{max}]$  
$ b  \epsilon  [b^{min}; b^{max}]$  

Contohnya:  
$ \max Z = c_1x_1 + c_2x_2 $  
s.t.  
$ a_{11}x_1 \ge b_1 $  
$ a_{22}x_2 \le b_2 $  
$ a_{31}x_1 + a_{32}x_2 \le b_3 $  
$ x_1, x_2 \ge 0 $  

Nilai parameternya adalah sebagai berikut:  

|Parameter|Rentang|
|:---:|:---:|
|$c_1$| 0.8 - 1.2|
|$c_2$| 1.5 - 2.5|
|$a_{11}$| 0.9 - 1.1|
|$a_{22}$| 0.7 - 1.3|
|$a_{31}$| 2.9 - 3.1|
|$a_{32}$| 3 - 5|
|$b_1$| 5.7 - 6.3|
|$b_2$| 3.2 - 4.8|
|$b_3$| 21.7- 25.3|

Dengan data di tabel di atas, maka formulasinya adalah:  
$ \max Z = 0.8x_1 + 1.5x_2 $  

s.t.  
$ 0.9x_1 \ge 6.3 $  
$ 1.3x_2 \le 3.2 $  
$ 3.1x_1 + 5x_2 \le 21.7 $  
$ x_1, x_2 \ge 0 $  

In [18]:
from pyomo import *

# untuk model concrete
m = ConcreteModel()

# definisikan variabel bebas
m.x1 = Var(within=NonNegativeReals)
m.x2 = Var(within=NonNegativeReals)

# definisikan fungsi objektif
m.objective = Objective(
                    expr = 0.8*m.x1+1.5*m.x2,
                    sense = maximize)

# definisikan constraint
m.constraint1 = Constraint(expr = 0.9*m.x1 >= 6.3)
m.constraint2 = Constraint(expr = 1.3*m.x2 <= 3.2)
m.constraint3 = Constraint(expr = 3.1*m.x1+5*m.x2 <= 21.7)

solver = SolverFactory('glpk')
hasil = solver.solve(m)

In [25]:
print(f"Nilai x1 adalah ", m.x1.value)
print(f"Nilai x2 adalah ", m.x2.value)
print(f"Nilai fungsi objektif adalah ", m.objective())

Nilai x1 adalah  7.0
Nilai x2 adalah  0.0
Nilai fungsi objektif adalah  5.6000000000000005


[Kembali ke daftar isi](#00_Daftar_Isi)  

<a id='09_Dynamic_Optimization'></a>
### 09. Dynamic Optimization
Aplikasi optimisasi matematika ke dalam suatu sistem dinamik disebut sebagai _dynamic optimization_.   

[Kembali ke daftar isi](#00_Daftar_Isi)